src/com/facebook/buck/jvm/kotlin/JarBackedReflectedKotlinc.java

```diff
@@ -0,0 +1,211 @@
+  private static final String EXIT_CODE_CLASS = "org.jetbrains.kotlin.cli.common.ExitCode";
+  private static final KotlincVersion VERSION = KotlincVersion.of("in memory");
+
+  private static final Function<Path, URL> PATH_TO_URL = p -> {
+    try {
+      return p.toUri().toURL();
+    } catch (MalformedURLException e) {
+      throw new RuntimeException(e);
+    }
+  };
+
+  // Used to hang onto the KotlinDaemonShim for the life of the buckd process
+  private static final Map<Set<String>, Object> kotlinShims =
+      new ConcurrentHashMap<>();
+
+  private final ImmutableSet<Path> compilerClassPath;
+
+  JarBackedReflectedKotlinc(
+      ImmutableSet<Path> compilerClassPath) {
+    this.compilerClassPath = compilerClassPath;
+  }
+
+  @Override
+  public KotlincVersion getVersion() {
+    return VERSION;
+  }
+
+  @Override
+  public String getDescription(
+      ImmutableList<String> options,
+      ImmutableSortedSet<Path> javaSourceFilePaths,
+      Path pathToSrcsList) {
+    StringBuilder builder = new StringBuilder("kotlinc ");
+    Joiner.on(" ").appendTo(builder, options);
+    builder.append(" ");
+    builder.append("@").append(pathToSrcsList);
+
+    return builder.toString();
+  }
+
+  @Override
+  public String getShortName() {
+    return "kotlinc";
+  }
+
+  @Override
+  public ImmutableList<String> getCommandPrefix(SourcePathResolver resolver) {
+    throw new UnsupportedOperationException("In memory kotlinc may not be used externally");
+  }
+
+  @Override
+  public int buildWithClasspath(
+      ExecutionContext context,
+      BuildTarget invokingRule,
+      ImmutableList<String> options,
+      ImmutableSortedSet<Path> kotlinSourceFilePaths,
+      Path pathToSrcsList,
+      Optional<Path> workingDirectory,
+      ProjectFilesystem projectFilesystem)
+      throws InterruptedException {
+
+    ImmutableList<String> args = ImmutableList.<String>builder()
+        .addAll(options)
+        .addAll(transform(
+            kotlinSourceFilePaths,
+            path -> projectFilesystem.resolve(path).toAbsolutePath().toString())
+        )
+        .build();
+
+    Set<File> compilerIdPaths = compilerClassPath
+        .stream()
+        .map(Path::toFile)
+        .collect(Collectors.toSet());
+
+    try {
+       Object compilerShim = kotlinShims.computeIfAbsent(
+          compilerIdPaths
+              .stream()
+              .map(File::getAbsolutePath)
+              .collect(Collectors.toSet()),
+          k -> loadCompilerShim(context));
+
+      Method compile = compilerShim
+          .getClass()
+          .getMethod("exec", PrintStream.class, String[].class);
+
+      Class<?> exitCodeClass = compilerShim
+          .getClass()
+          .getClassLoader()
+          .loadClass(EXIT_CODE_CLASS);
+
+      Method getCode = exitCodeClass.getMethod("getCode");
+
+      Object exitCode = compile.invoke(
+          compilerShim,
+          new UncloseablePrintStream(context.getStdErr()),
```

Comment:
@brettwooldridge an Infer run is complaining about a [resource leak](http://fbinfer.com/docs/infer-bug-types.html#RESOURCE_LEAK) here:  Any ideas on the best way to fix this?